### ISSUE TITLE
bazel hitting "This service does not support downloading partial files"

### ISSUE THREAD

**Issue by ailiev** [buildbarn/bb-storage#15]
I've been hitting this restriction occasionally, I think correlated with large/wide builds. Bazel 0.26.0. The buildbarn error is reported here https://github.com/buildbarn/bb-storage/blob/master/pkg/cas/byte_stream_server.go#L63

I don't know under what circumstances bazel would request a part of a file, but in my experience it's infrequent but not never.

Bazel error message:
 
```
(10:11:45) ERROR: …/BUILD.bazel:1:1: Extracting interface … failed (Exit 34). Note: Remote connection/protocol failed with: execution failed io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files: java.io.IOException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files
        at com.google.devtools.build.lib.remote.GrpcRemoteCache.lambda$downloadBlob$4(GrpcRemoteCache.java:318)
        at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:175)
        at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:162)
        at com.google.common.util.concurrent.AbstractCatchingFuture.run(AbstractCatchingFuture.java:107)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
        at com.google.common.util.concurrent.AbstractFuture.setFuture(AbstractFuture.java:752)
        at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.setResult(AbstractCatchingFuture.java:185)
        at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.setResult(AbstractCatchingFuture.java:162)
        at com.google.common.util.concurrent.AbstractCatchingFuture.run(AbstractCatchingFuture.java:116)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:711)
        at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:54)
        at com.google.devtools.build.lib.remote.GrpcRemoteCache$2.onError(GrpcRemoteCache.java:358)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:398)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files
        at io.grpc.Status.asRuntimeException(Status.java:532)
... 19 more
```

**Comment by abentley** [comment]
I have also experienced this.  It may be intermittent.

[473 / 638] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 5s remote ... (26 actions, 21 running)

[473 / 638] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 5s remote ... (26 actions, 21 running)

[647 / 1,310] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 7s remote ... (63 actions, 59 running)

[1,089 / 1,410] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 10s remote ... (64 actions, 62 running)

[1,218 / 1,460] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 14s remote ... (64 actions, 63 running)

[1,319 / 1,530] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 18s remote ... (81 actions, 63 running)

[1,380 / 1,572] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 23s remote ... (64 actions, 63 running)

[1,494 / 1,683] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 28s remote ... (64 actions, 63 running)

[1,619 / 1,806] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 35s remote ... (64 actions, 63 running)

[1,719 / 1,902] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 42s remote ... (64 actions, 63 running)

[1,813 / 2,001] SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar; 51s remote ... (64 actions, 63 running)

ERROR: /root/.cache/bazel/_bazel_root/f491cbe24ef733ba69e71fc89f5d8689/external/bazel_tools/tools/jdk/BUILD:306:1: SkylarkAction external/bazel_tools/tools/jdk/platformclasspath.jar failed (Exit 34). Note: Remote connection/protocol failed with: execution failed io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files: java.io.IOException: io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files

	at com.google.devtools.build.lib.remote.GrpcRemoteCache.lambda$downloadBlob$4(GrpcRemoteCache.java:318)

	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:175)

	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:162)

	at com.google.common.util.concurrent.AbstractCatchingFuture.run(AbstractCatchingFuture.java:107)

	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)

	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)

	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)

	at com.google.common.util.concurrent.AbstractFuture.setFuture(AbstractFuture.java:752)

	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.setResult(AbstractCatchingFuture.java:185)

	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.setResult(AbstractCatchingFuture.java:162)

	at com.google.common.util.concurrent.AbstractCatchingFuture.run(AbstractCatchingFuture.java:116)

	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)

	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)

	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)

	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:711)

	at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:54)

	at com.google.devtools.build.lib.remote.GrpcRemoteCache$2.onError(GrpcRemoteCache.java:358)

	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434)

	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)

	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)

	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)

	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)

	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)

	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)

	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)

	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:398)

	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)

	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)

	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)

	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)

	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)

	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)

	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)

	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

	at java.base/java.lang.Thread.run(Unknown Source)

Caused by: io.grpc.StatusRuntimeException: UNIMPLEMENTED: This service does not support downloading partial files

	at io.grpc.Status.asRuntimeException(Status.java:532)

	... 19 more

[1,962 / 2,121] Executing genrule //third_party/ffmpeg:extract; 50s remote

**Comment by ailiev** [comment]
Some poking around bazel shows that it does a partial read with an offset and no limit (ie read to the end) if a previous read of the same blob returned only a prefix of the contents and then had some failure.

**Comment by EdSchouten** [comment]
Fixed in #34! bb-storage now contains [a buffering layer](https://godoc.org/github.com/buildbarn/bb-storage/pkg/blobstore/buffer) that supports offset reads natively (search for `ToChunkReader()`). Thanks for reporting this issue!
